### DISCUSSION TITLE
How to read nested objects in json

### DISCUSSION THREAD

**Discussion by newjsonuser** [nlohmann/json#2758] [Q&A] [answered]
Hi, I want to read the nested objects for the below example json file. I do not understand how to get CandidateArray[0].name which should be Matt. I am new to reading json files.

```json
 {
    "Company": {
        "Employees": [
            {
                "candidate1": {
                    "name": "John",
                    "degree": "xx",
                    "enrollment": "2019",
                    "id": 12345
                },
                "candidate2": {
                    "name": "Joe",
                    "degree": "xx",
                    "enrollment": "2019",
                    "id": 123456
                }
            },
            {
              "CandidateArray": [
                {
                  "name": "Matt",
                  "degree": "xx",
                  "enrollment": "2009",
                  "id": 1234,
                  "experience": 10
                },
                {
                  "name": "Paul",
                  "degree": "xx",
                  "enrollment": "2009",
                  "id": 1235,
                  "experience": 10
                }
              ]
            }
        ]
    }
}
```

```cpp
#include <iostream>
#include <filesystem>
#include <fstream>
#include "json.hpp"

std::string findRecursive(const nlohmann::json& j, const std::string& path)
{
    std::string name;
    if (j.is_object() || j.is_array())
    {
        for (auto &element : j.items())
        {
            std::cout << "Name: " << element.key() << "\n";

            if ((element.value().is_object() || element.value().is_array()))
            {
                name = findRecursive(element.value(), path);
            }
            else
            {
                std::string elemString = element.key();
                if (elemString._Equal("name"))
                {
                    std::cout << element.value() << std::endl;
                    name = element.value();
                    break;
                }
            }
        }
    }
    return name;
}
int main()
{
    using json = nlohmann::json;

    std::ifstream file1("testjson.json");
    if (!file1)
    {
        return -1;
    }

    json j;
    file1 >> j;

    std::string name = findRecursive(j, "Company/Employees/CandidateArray/name");
    std::cout << name << std::endl;

return 0;
}
```

I had a recursive function but it returns name "Paul". How do I access any element (example how do I make it return "Matt")?Also is there an efficient way to access all values (like name, id at the same time). Thanks for suggestions and help.

**Comment by nlohmann** [comment, marked as answer]
Maybe you can use JSON Pointers, see https://github.com/nlohmann/json#json-pointer-and-json-patch.

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?

**Reply by newjsonuser** [reply]
No, I am good. Thanks

**Reply by PreetM07** [reply]
Hi... 
@newjsonuser has JSON pointer solved ur issue?
I can see that when I try to travers deep it throws exception...

terminate called after throwing an instance of 'nlohmann::detail::parse_error'
  what():  [json.exception.parse_error.109] parse error: array index 'source' is not a number

I have tried cout << j_complete2["/items/source/text"_json_pointer];

Any help?

**Reply by nlohmann** [reply]
The error indicates that whatever you addresses with `/items` is an array, and to address an element of this array, you need to pass a number. However, `source` was passed instead. So (without knowing how your JSON looks like), I would assume you need to try something like

```cpp
j_complete2["/items/0/source/text"_json_pointer]
```

**Reply by newjsonuser** [reply]
@PreetM07  I did not know how to use JSON pointers. I used recursive function to get the objects.

**Reply by PreetM07** [reply]
> The error indicates that whatever you addresses with `/items` is an array, and to address an element of this array, you need to pass a number. However, `source` was passed instead. So (without knowing how your JSON looks like), I would assume you need to try something like
> 
> ```c++
> j_complete2["/items/0/source/text"_json_pointer]
> ```

That's it. Thanks.

**Reply by PreetM07** [reply]
> @PreetM07 I did not know how to use JSON pointers. I used recursive function to get the objects.

would be helpful to someone (also to me :P) if you can provide the working recursion. I think this one is yet partial?

**Reply by PreetM07** [reply]
> The error indicates that whatever you addresses with `/items` is an array, and to address an element of this array, you need to pass a number. However, `source` was passed instead. So (without knowing how your JSON looks like), I would assume you need to try something like
> 
> ```c++
> j_complete2["/items/0/source/text"_json_pointer]
> ```

Now how should one iterate through this?
let's say I want to iterate...
"/items/0/source/text"_json_pointer
"/items/1/source/text"_json_pointer
"/items/1/source/text"_json_pointer
"/items/1/source/text"_json_pointer
"/items/1/source/text"_json_pointer...

string operators doesn't work with _json_pointer

I tried...
maf = "\"/items/" + to_string(i) + "source/text\"";  //also without quotes but seems like json_pointer is "" overloaded.
                nlohmann::json::json_pointer jp = nlohmann::json::json_pointer(std::string(maf, maf.length()));
                outputFileHandler << j_complete2[jp] << ",";

HELP?

**Reply by nlohmann** [reply]
I am note sure if you really want to use JSON Pointers in this case. Then a nested iteration like

```cpp
for (int i = 0; i < j_complete2["items"].size(); ++i)
{
    std::cout << j_complete2["items"][i]["source"]["text"] << '\n';
}
```

also works.

**Reply by PreetM07** [reply]
> I am note sure if you really want to use JSON Pointers in this case. Then a nested iteration like
> 
> ```c++
> for (int i = 0; i < j_complete2["items"].size(); ++i)
> {
>     std::cout << j_complete2["items"][i]["source"]["text"] << '\n';
> }
> ```
> 
> also works.

That also makes sense. Thank you very much. But would be great for someone's future reference, adding this trick using json_pointer here or in doc

**Reply by nlohmann** [reply]
I agree the documentation of JSON Pointers needs improvement. I will see what I can do.

**Reply by nlohmann** [reply]
I overworked the section https://json.nlohmann.me/features/json_pointer/.

**Reply by PreetM07** [reply]
Lovely documentation. Thank you so much.
